third_party/pcsc-lite/naclport/js_client/src/reader-tracker-through-pcsc-api.js

```diff
@@ -0,0 +1,435 @@
+    result.get(
+        promiseResolver.resolve.bind(promiseResolver),
+        function(errorCode) {
+          if (errorCode == API.SCARD_E_NO_READERS_AVAILABLE) {
+            promiseResolver.resolve([]);
+          } else {
+            promiseResolver.reject(new Error(
+                'Failed to get the list of readers from PC/SC with error ' +
+                'code ' + GSC.DebugDump.dump(errorCode)));
+          }
+        });
+  }, function(error) {
+    promiseResolver.reject(new Error(
+        'Failed to get the list of readers from PC/SC: ' + error));
+  });
+
+  return promiseResolver.promise;
+};
+
+/**
+ * Makes a promise of reader states that are reported currently by PC/SC.
+ * @param {!API} api
+ * @param {!API.SCARDCONTEXT} sCardContext
+ * @param {!Array.<string>} readerNames
+ * @return {!goog.Promise.<!Array.<!API.SCARD_READERSTATE_OUT>|null>} Either the
+ * states of the readers, or the null value in case of intermittent error.
+ * @private
+ */
+ReaderTrackerThroughPcscApi.prototype.makeReaderStatesPromise_ = function(
+    api, sCardContext, readerNames) {
+  /** @type {!goog.promise.Resolver.<!Array.<!API.SCARD_READERSTATE_OUT>|null>} */
+  const promiseResolver = goog.Promise.withResolver();
+
+  const readerStatesIn = goog.array.map(readerNames, function(readerName) {
+    return new API.SCARD_READERSTATE_IN(readerName, API.SCARD_STATE_UNAWARE);
+  });
+
+  api.SCardGetStatusChange(sCardContext, 0, readerStatesIn).then(
+      function(result) {
+        result.get(
+            promiseResolver.resolve.bind(promiseResolver),
+            function(errorCode) {
+              if (errorCode == API.SCARD_E_UNKNOWN_READER) {
+                this.logger_.warning(
+                    'Getting the statuses of the readers from PC/SC finished ' +
+                    'unsuccessfully due to removal of the tracked reader. A ' +
+                    'retry will be attempted after some delay');
+                promiseResolver.resolve(null);
+              } else {
+                promiseResolver.reject(new Error(
+                    'Failed to get the reader statuses from PC/SC with error ' +
+                    'code ' + GSC.DebugDump.dump(errorCode)));
+              }
+            },
+            this);
+      }, function(error) {
+        promiseResolver.reject(new Error(
+            'Failed to get the reader statuses from PC/SC: ' + error));
+      });
+
+  return promiseResolver.promise;
+};
+
+/**
+ * Makes a promise that will be resolved once some change with the specified
+ * state is detected, or when the timeout passes.
+ *
+ * This also watches for the reader adding/removing events, but this,
+ * unfortunately, cannot be done in a race-free manner: the PC/SC
+ * SCardGetStatusChange function would notify only about those reader
+ * adding/removing events that occurred during its call. So the reader
+ * adding/removing may happen.
+ *
+ * FIXME(emaxx): Think whether it makes sense to make the implementation more
```

Comment:
nit: TODO(issue #125) - or just delete this since it is mentioned above already.

Reply:
Done (removed)